clients/src/main/java/org/apache/kafka/clients/consumer/internals/ListOffsetsRequestManager.java

```diff
@@ -109,38 +109,42 @@ public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
         return pollResult;
     }
 
+
     /**
      * Retrieve offsets for the given partitions and timestamp.
      *
-     * @param partitions        Partitions to get offsets for
-     * @param timestamp         Target time to look offsets for
-     * @param requireTimestamps True if this should fail with an UnsupportedVersionException if
-     *                          the broker does not support fetching precise timestamps for offsets
-     * @return Future containing the map of offsets retrieved for each partition. The future will
-     * complete when the responses for the requests are received and processed following a call
-     * to {@link #poll(long)}
+     * @param timestampsToSearch Partitions and target timestamps to get offsets for
+     * @return Future containing the map of {@link TopicPartition} and {@link OffsetAndTimestamp}
+     * found (offset of the first message whose timestamp is greater than or equals to the target
+     * timestamp).The future will complete when the requests responses are received and
+     * processed, following a call to {@link #poll(long)}
      */
-    public CompletableFuture<Map<TopicPartition, Long>> fetchOffsets(final Set<TopicPartition> partitions,
-                                                                     final long timestamp,
-                                                                     final boolean requireTimestamps) {
-        metadata.addTransientTopics(offsetFetcherUtils.topicsForPartitions(partitions));
+    public CompletableFuture<Map<TopicPartition, OffsetAndTimestamp>> fetchOffsets(
+            final Map<TopicPartition, Long> timestampsToSearch,
+            final boolean requireTimestamps) {
+        if (timestampsToSearch.isEmpty()) {
+            return CompletableFuture.completedFuture(Collections.emptyMap());
+        }
+        metadata.addTransientTopics(offsetFetcherUtils.topicsForPartitions(timestampsToSearch.keySet()));
 
-        Map<TopicPartition, Long> timestampsToSearch = partitions.stream()
-                .collect(Collectors.toMap(Function.identity(), tp -> timestamp));
         ListOffsetsRequestState listOffsetsRequestState = new ListOffsetsRequestState(
                 timestampsToSearch,
                 requireTimestamps,
                 offsetFetcherUtils,
                 isolationLevel);
         listOffsetsRequestState.globalResult.whenComplete((result, error) -> {
             metadata.clearTransientTopics();
-            log.debug("Fetch offsets completed for partitions {} and timestamp {}. Result {}, " +
-                    "error", partitions, timestamp, result, error);
+            log.debug("Fetch offsets completed for partitions and timestamps {}. Result {}, " +
```

Comment:
Will the user see the string `error` in the logs even when the request is complete?